kytos/core/buffers/buffers.py

```diff
@@ -123,53 +125,28 @@ def full(self):
         return self._queue.sync_q.full()
 
 
-class KytosBuffers:
-    """Set of KytosEventBuffer used in Kytos."""
-
-    def __init__(self):
-        """Build four KytosEventBuffers.
-
-        :attr:`conn`: :class:`~kytos.core.buffers.KytosEventBuffer` with events
-        received from connection events.
-
-        :attr:`raw`: :class:`~kytos.core.buffers.KytosEventBuffer` with events
-        received from network.
+class RateLimitedBuffer(KytosEventBuffer):
+    """
+    Extension of KytosEventBuffer with ratelimiting capabilities.
+    """
+    def __init__(
+        self, *args,
+        strategy: lstrategies.RateLimiter,
+        limit: limits.RateLimitItem,
+        gen_identifiers: Callable[[KytosEvent], Iterable[Hashable]],
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self.strategy = strategy
+        self.limit = limit
+        self.gen_identifiers = gen_identifiers
 
-        :attr:`msg_in`: :class:`~kytos.core.buffers.KytosEventBuffer` with
-        events to be received.
-
-        :attr:`msg_out`: :class:`~kytos.core.buffers.KytosEventBuffer` with
-        events to be sent.
-
-        :attr:`app`: :class:`~kytos.core.buffers.KytosEventBuffer` with events
-        sent to NApps.
-        """
-        self._pool_max_workers = get_thread_pool_max_workers()
-        self.conn = KytosEventBuffer("conn")
-        self.raw = KytosEventBuffer("raw", maxsize=self._get_maxsize("sb"))
-        self.msg_in = KytosEventBuffer("msg_in",
-                                       maxsize=self._get_maxsize("sb"),
-                                       queue_cls=PriorityQueue)
-        self.msg_out = KytosEventBuffer("msg_out",
-                                        maxsize=self._get_maxsize("sb"),
-                                        queue_cls=PriorityQueue)
-        self.app = KytosEventBuffer("app", maxsize=self._get_maxsize("app"))
-
-    def get_all_buffers(self):
-        """Get all KytosEventBuffer instances."""
-        return [
-            event_buffer for event_buffer in self.__dict__.values()
-            if isinstance(event_buffer, KytosEventBuffer)
-        ]
-
-    def _get_maxsize(self, queue_name):
-        """Get queue maxsize if it's been set."""
-        return self._pool_max_workers.get(queue_name, 0)
-
-    def send_stop_signal(self):
-        """Send a ``kytos/core.shutdown`` event to each buffer."""
-        LOG.info('Stop signal received by Kytos buffers.')
-        LOG.info('Sending KytosShutdownEvent to all apps.')
-        event = KytosEvent(name='kytos/core.shutdown')
-        for buffer in self.get_all_buffers():
-            buffer.put(event)
+    async def aget(self):
```

Comment:
David, another requirement to consider:

- Also, rate limit at the `put` and `aput` level so before actually putting in the queue, essentially rate limiting as close to the source as possible as far as the queue is concerned. This also implies supporting both sync and async for this end of the queue.

Let me elaborate on the use case in the next comments.

Reply:
This is partly related [a prior chaining discussion](https://github.com/kytos-ng/kytos/pull/412#discussion_r1332156924), but here it would only chain on the `put|aput` where any publisher could register a particular rate limiter that it'll use. Or alternatively provide a context manager or something similar where publisher who need to be rate limiter before putting in the queue they can use that context or async context, although I believe the former would be less boilerplate and more ergonomic to use.

Considering that we've been using bounded queues, this will also contribute to a bit more of fairness with other concurrent producers who might be sending other type of messages regardless of using a priority queue or not (and even more when it's using a priority queue), when rate limiters are also being used on the `put|aput` side. It's a subtle difference, but if the queue is properly sized for the current number of events going on, then it can make a significant difference since the queue won't be constantly getting full.

`````
                                                          producer 1 (put)
core event_handler aget <---  Queue X   <----  put|aput   producer 2 (aput)
                                                          ....
                                                          producer n (put|aput)
    |
    |
    \/
________________________________

NApps event handlers run by

thread pool and/or asyncio task

________________________________


`````

`aget` working as a global rate limiter with a given identifier as it's currently implemented, and then also structure on `put|aput` in a way where in the future it's also possible to throttle with a given unique id context, allowing it to essentially replacing implementing rate limiting at the producer level as we've been doing for `mef_eline` and `telemetry_int` (even though the global `aget` rate limit will already solve a major part of the problem).

Reply:
Concrete use case example with our existing Kytos-ng NApps:

- globally rate limit any messages per `'destination.id'` on `msg_out` at `100/second` AND individually rate limit any flows from `mef_eline` at `'30/sec'` and any from `telemetry_int` at `'30/sec'`. 

On ``msg_out``, `flow_manager` is responsible for pushing FlowMods, so it's centralizing the events, so on `flow_manager` it could on behalf of its clients rate limit based on the flow [`owner`](https://github.com/kytos-ng/flow_manager/blob/master/db/models/__init__.py#L109) value:

```
                                                     mef_eline (via requests and KytosEvent)
aget  <---  Queue msg_out <---  put   flow_manager

                                                    telemetry_int (via KytosEvent)
```

Do you see what I mean? 

Can you also support this or propose another approach to also solve the publisher side rate limit?